tests/test_validation.py

```diff
@@ -208,3 +206,296 @@ def test_validation_on_write(tmp_path):
 
     with pytest.warns(ValidationWarning):
         model.save(file_path)
+
+def test_validate_on_assignment_default():
+    model = ValidationModel()
+    assert model._validate_on_assignment==True
+
+    with pytest.warns(ValidationWarning):
+        model.meta.string_attribute = 42  # Bad assignment
+    assert model.meta.string_attribute is None
+
+
+def test_validate_on_assignment_false(tmp_path):
+    file_path = tmp_path/"test.asdf"
+
+    model = ValidationModel(validate_on_assignment=False)
+    assert model._validate_on_assignment==False
+
+    with pytest.warns(None) as warnings:
+        model.meta.string_attribute = 42  # Bad assignment should cause no warning
+    assert len(warnings) == 0
+    assert model.meta.string_attribute == 42
+
+    with pytest.warns(ValidationWarning):
+        model.save(file_path)
```

Comment:
Looking through the tests more carefully, I see there's nothing unique about writing out an invalid tree when `validate_on_assignment=True`.  An invalid tree is an invalid tree, so `test_validation_on_write` test above should be sufficient to test this.  Was good to verify this while writing tests, but in terms of the final test code, I think this test case is covered.

Reply:
I merged your PR to my branch.  I liked the way your refactored my tests.